receiver/k8sclusterreceiver/watcher.go

```diff
@@ -130,6 +147,22 @@ func (rw *resourceWatcher) onUpdate(oldObj, newObj interface{}) {
 	rw.syncMetadataUpdate(oldMetadata, newMetadata)
 }
 
+// waitForInitialInformerSync waits for initial cache sync, after the informers
+// are started up. This is achieved by resourceWatcher.informersHaveSynced. This
+// method is intended to be used by the ResourceEventHandler callbacks to ensure
+// that data sync does not start until the cache is synced as the events could be
+// received by the handler even before the informer caches being synced.
+func (rw *resourceWatcher) waitForInitialInformerSync() {
+	// Return immediately if already synced.
+	if rw.informersHaveSynced.Load() {
+		return
+	}
+
+	for !rw.informersHaveSynced.Load() {
```

Comment:
This can keep running infinitely. Should we have a timeout to avoid getting stuck forever if something goes wrong?

Reply:
I originally had added a timeout here. But then later removed it since we want to have ensure the we have all the data before the sync begins. Also, playing around with this, so far, it looks like the caches are synced in < 20s for our lab deployment (however, might vary based on size of the deployment). We could have a timeout here to safeguard against the risk of running infinitely, but that would potentially mean we sync incomplete metadata during the initial sync. On subsequent metadata updates to the relevant dimension (or label in OTel), if the cache is synced, the missed out metadata will also be synced but until then the synced metadata would be incomplete. @tigrannajaryan @dmitryax

Reply:
I don't think we want to have infinite loops that are not cancellable in the Collector. This will create problems for graceful shutdown in the future too. We should probably add cancellation support at least.

Reply:
I agree, will make an update to handle this better.

Reply:
@asuresh4 correct me if I'm wrong here, but as I understood the original timeout was implemented in a "best effort" way. If cache sync is not complete we ignore parts of the metadata that were not able to sync. That actually causes corrupted metadata updates in the current version of collector. We should avoid it, but I'm not sure what's the best way. Probably better to fail to start the receiver. But if we don't do anything it'll be handled by liveness probes anyway.

Reply:
I'd suggest the following: 
1. make the cache sync not blocking the pipeline building (not sure if it is already);
2. add cancellation support;
3. have a reasonable default timeout for cache sync, like 2-5 minutes;
2. by timeout, print an error message and deactivate the receiver.
Not sure how feasible is this. @tigrannajaryan @asuresh4 what do you think?

Reply:
Discussed offline, accepting @dmitryax's suggestion, except in step 4 we call ReportFatalError and terminate the Collector since we believe this is critical because there is only one instance of Collector that is responsible for this collection.

Reply:
Updated as per our discussion. I've set the timeout to be 10 minutes (which I believe should be long enough for all cases), after which the receiver reports a fatal error. Also, leveraged cancellation support of `rw.sharedInformerFactory.WaitForCacheSync` to make sure it returns if the receiver is Shutdown while it's still waiting for the sync. PTAL @dmitryax @tigrannajaryan.